core/src/main/scala/kafka/server/OffsetCheckpoint.scala

```diff
@@ -16,95 +16,99 @@
  */
 package kafka.server
 
+import java.nio.file.Paths
+import java.util.regex.Pattern
+
+import org.apache.kafka.common.utils.Utils
+
 import scala.collection._
 import kafka.utils.Logging
 import kafka.common._
 import java.io._
 
+object OffsetCheckpoint {
+  private val WhiteSpacesPattern = Pattern.compile("\\s+")
+  private val CurrentVersion = 0
+}
+
 /**
  * This class saves out a map of topic/partition=>offsets to a file
  */
 class OffsetCheckpoint(val file: File) extends Logging {
+  import OffsetCheckpoint._
+  private val path = file.toPath.toAbsolutePath
+  private val tempPath = Paths.get(path.toString + ".tmp")
   private val lock = new Object()
-  new File(file + ".tmp").delete() // try to delete any existing temp files for cleanliness
   file.createNewFile() // in case the file doesn't exist
 
   def write(offsets: Map[TopicAndPartition, Long]) {
     lock synchronized {
       // write to temp file and then swap with the existing file
-      val temp = new File(file.getAbsolutePath + ".tmp")
-
-      val fileOutputStream = new FileOutputStream(temp)
+      val fileOutputStream = new FileOutputStream(tempPath.toFile)
       val writer = new BufferedWriter(new OutputStreamWriter(fileOutputStream))
       try {
-        // write the current version
-        writer.write(0.toString)
+        writer.write(CurrentVersion.toString)
         writer.newLine()
-      
-        // write the number of entries
+
         writer.write(offsets.size.toString)
         writer.newLine()
 
-        // write the entries
         offsets.foreach { case (topicPart, offset) =>
-          writer.write("%s %d %d".format(topicPart.topic, topicPart.partition, offset))
+          writer.write(s"${topicPart.topic} ${topicPart.partition} $offset")
           writer.newLine()
         }
-      
-        // flush the buffer and then fsync the underlying file
+
         writer.flush()
         fileOutputStream.getFD().sync()
       } finally {
         writer.close()
       }
-      
-      // swap new offset checkpoint file with previous one
-      if(!temp.renameTo(file)) {
-        // renameTo() fails on Windows if the destination file exists.
-        file.delete()
-        if(!temp.renameTo(file))
-          throw new IOException("File rename from %s to %s failed.".format(temp.getAbsolutePath, file.getAbsolutePath))
-      }
+
+      Utils.atomicMoveWithFallback(tempPath, path)
```

Comment:
I wondered the same thing as you and I don't have a Windows machine to test either. Googling seems to indicate that the new atomic move API actually works on Windows too if the destination file exists (unlike File.rename), but it's based on other people saying they've tried it. I added the fallback just in case. The UNIX implementation does a delete followed by a move and I'd expect the Windows one to do the same. I'll actually check the code and let you know.